homeassistant/components/__init__.py

```diff
@@ -143,9 +143,11 @@ def async_handle_turn_service(service):
         yield from asyncio.wait(tasks, loop=hass.loop)
 
     hass.services.async_register(
-        ha.DOMAIN, SERVICE_TURN_OFF, async_handle_turn_service)
+        ha.DOMAIN, SERVICE_TURN_OFF, async_handle_turn_service,
+        state_to_set=STATE_OFF)
     hass.services.async_register(
-        ha.DOMAIN, SERVICE_TURN_ON, async_handle_turn_service)
+        ha.DOMAIN, SERVICE_TURN_ON, async_handle_turn_service,
+        state_to_set=STATE_ON)
```

Comment:
We could have a function in each component that can be used to lookup the component specific services and states. The function could be bound to `hass`. But we want to lookup by domain and state once, and then get the correct result. From the state we can't know what component is involved, only the domain, right? I don't see how we would combine calls to all the functions on all components under a domain, easily?

I'm still thinking registering the states together with the services is the way to go. If you see a way of solving the lookup for all components under a domain, we could follow up on that.